pkg/experiment/metastore/index/index.go

```diff
@@ -0,0 +1,462 @@
+package index
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/oklog/ulid"
+	"github.com/prometheus/common/model"
+	"golang.org/x/sync/errgroup"
+
+	metastorev1 "github.com/grafana/pyroscope/api/gen/proto/go/metastore/v1"
+)
+
+type Index struct {
+	Config *Config
+
+	partitionMu      sync.Mutex
+	loadedPartitions *lru.Cache[PartitionKey, *indexPartition]
```

Comment:
If you mean using partition+tenant as the key, that will also have interference (busy tenants would push others out of the cache). For separation we could do separate caches per tenant (complex, costly) or go back to the previous solution (an unbounded TTL-based cache, uneven memory usage).

As discussed separately, this cache is also not suitable because we can get the active (for writes) partition unloaded with a large query. This again can be solved by switching to a different caching strategy (ARC, 2Q, etc.).

Personally, given our usage patterns (frequent writes, infrequent reads) I lean towards a custom solution similar to what we had before, with an upper bound on how many items stay in memory and explicit checks that prevent the write partition to be unloaded.

Reply:
The cache is now per partition+tenant, though I've switched to a custom caching solution where each tenant gets `Config.PartitionCacheSize` slots in the cache.